Bruce/nightly agent3 (#26107)
ray-project/ray: doc/source/serve/production.md

### Using a Remote Cluster

By default, `serve deploy` deploys to a cluster running locally. However, you should also use `serve deploy` whenever you want to deploy your Serve application to a remote cluster. `serve deploy` takes in an optional `--address/-a` argument where you can specify your remote Ray cluster's dashboard agent address. This address should be of the form:

```
[YOUR_RAY_CLUSTER_URI]:[DASHBOARD AGENT PORT]
```

As an example, the address for the local cluster started by `ray start --head` is `http://127.0.0.1:52365`. We can explicitly deploy to this address using the command

```console
$ serve deploy config_file.yaml -a http://127.0.0.1:52365
```

> The Ray dashboard agent's default port is 52365. This port may be different if:

Reviewer: Since there's only one reason why it might be different, can we remove the bullet point syntax? We can say

```markdown
"The Ray dashboard agent's default port is 52365. You can set it to a different value using the `--dashboard-agent-listen-port` argument when running `ray start`."
```

* You explicitly set it using the `--dashboard-agent-listen-port` argument when running `ray start`.

:::{note}

> If the port 52365 is unavailable when Ray starts, the dashboard agent’s HTTP server will fail. However, the dashboard agent and Ray will continue to run.

Reviewer: ```suggestion
If the port 52365 (or whichever port you specify with `--dashboard-agent-listen-port`) is unavailable when Ray starts, the dashboard agent’s HTTP server will fail. However, the dashboard agent and Ray will continue to run.
```

> You can check if an agent’s HTTP server is running by sending a curl request like: `curl http://{node_ip}:{dashboard agent port}/api/serve/deployments/`. If the request succeeds, the server is running on that node. If the request fails, the server is not running on that node. To launch the server on that node, terminate the process occupying the dashboard agent’s port, and restart Ray on that node.

Reviewer: ```suggestion
You can check if an agent’s HTTP server is running by sending a curl request: `curl http://{node_ip}:{dashboard_agent_port}/api/serve/deployments/`. If the request succeeds, the server is running on that node. If the request fails, the server is not running on that node. To launch the server on that node, terminate the process occupying the dashboard agent’s port, and restart Ray on that node.
```

:::

:::{tip}

> By default, all the Serve CLI commands assume that you're working with a local cluster, they use the Ray agent address associated with a local cluster started by `ray start --head`. However, if the `RAY_AGENT_ADDRESS` environment variable is set, all Serve CLI commands will default to that value instead.

Reviewer: ```suggestion
By default, all the Serve CLI commands assume that you're working with a local cluster. All Serve CLI commands, except `serve start` and `serve run` use the Ray agent address associated with a local cluster started by `ray start --head`. However, if the `RAY_AGENT_ADDRESS` environment variable is set, these Serve CLI commands will default to that value instead.

Similarly, `serve start` and `serve run`, use the Ray head node address associated with a local cluster by default. If the `RAY_ADDRESS` environment variable is set, they will use that value instead.
```

You can check if an agent’s HTTP server is running by sending a curl request: `curl http://{node_ip}:{dashboard_agent_port}/api/serve/deployments/`. If the request succeeds, the server is running on that node. If the request fails, the server is not running on that node. To launch the server on that node, terminate the process occupying the dashboard agent’s port, and restart Ray on that node.
:::

:::{tip}
By default, all the Serve CLI commands assume that you're working with a local cluster. All Serve CLI commands, except `serve start` and `serve run` use the Ray agent address associated with a local cluster started by `ray start --head`. However, if the `RAY_AGENT_ADDRESS` environment variable is set, these Serve CLI commands will default to that value instead.
	
Similarly, `serve start` and `serve run`, use the Ray head node address associated with a local cluster by default. If the `RAY_ADDRESS` environment variable is set, they will use that value instead.

> You can check this variable's value by running:

Reviewer: ```suggestion
You can check `RAY_AGENT_ADDRESS`'s value by running:
```

By default, `serve deploy` deploys to a cluster running locally. However, you should also use `serve deploy` whenever you want to deploy your Serve application to a remote cluster. `serve deploy` takes in an optional `--address/-a` argument where you can specify the dashboard address of your remote Ray cluster. This address should be of the form:

```
[YOUR_RAY_CLUSTER_URI]:[DASHBOARD AGENT PORT]
```

As an example, the address for the local cluster started by `ray start --head` is `http://127.0.0.1:52365`. We can explicitly deploy to this address using the command

```console
$ serve deploy config_file.yaml -a http://127.0.0.1:52365
```

The Ray dashboard agent's default port is 52365. This port may be different if:
* You explicitly set it using the `--dashboard-agent-listen-port` argument when running `ray start`.

> * Port 52365 was unavailable when Ray started. In that case, the dashboard agent port is randomly generated.

Reviewer: Is this still true? I think now we crash if port is unavailable.

If that's the case, I would remove this bullet point and replace it with a sentence mentioning that if the port was occupied before the agent started, the agent will still run but be unable to respond to Serve CLI requests or REST API requests. I would also mention (1) how the user should check if that's the case (what commands they should run to see if the port is occupied) and (2) what the user should do to make the Serve CLI work again.

The Serve CLI offers two commands to help you inspect your Serve application in production: `serve config` and `serve status`.

If you're working with a remote cluster, `serve config` and `serve status` also offer an `--address/-a` argument to access your cluster.

> By default, all the Serve CLI commands assume that you're working with a local cluster, so if you don't specify an `--address/-a` value, they use the Ray agent address associated with a local cluster started by `ray start --head`. However, if the `RAY_AGENT_ADDRESS` environment variable is set, all Serve CLI commands will default to that value instead (unless you also specify an `--address/-a` value).

Reviewer: Can we include a link to the `(serve-in-production-remote-cluster)` section somewhere in this paragraph, so users can refer to the extensive explanation?

The Serve CLI offers two commands to help you inspect your Serve application in production: `serve config` and `serve status`.

If you're working with a remote cluster, `serve config` and `serve status` also offer an `--address/-a` argument to access your cluster.

> By default, all the Serve CLI commands assume that you're working with a local cluster, so if you don't specify an `--address/-a` value, they use the Ray agent address associated with a local cluster started by `ray start --head`. However, if the `RAY_AGENT_ADDRESS` environment variable is set, all Serve CLI commands will default to that value instead (unless you also specify an `--address/-a` value). Check out [the previous section](serve-in-production-remote-cluster) for more info on this argument.

Reviewer: I think this paragraph is a bit redundant with the info from the previous section. I think we can revert this part of the doc change and link to the previous section instead.
